drivers/counter/ds3231.c

```diff
@@ -0,0 +1,1368 @@
+/*
+ * Copyright (c) 2019-2020 Peter Bigot Consulting, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT maxim_ds3231
+
+#include <device.h>
+#include <drivers/rtc/ds3231.h>
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <kernel.h>
+#include <logging/log.h>
+#include <sys/timeutil.h>
+#include <sys/util.h>
+
+LOG_MODULE_REGISTER(DS3231, CONFIG_COUNTER_LOG_LEVEL);
+
+#define REG_MONCEN_CENTURY 0x80
+#define REG_HOURS_12H 0x40
+#define REG_HOURS_PM 0x20
+#define REG_HOURS_20 0x20
+#define REG_HOURS_10 0x20
+#define REG_DAYDATE_DOW 0x40
+#define REG_ALARM_IGN 0x80
+
+enum {
+	SYNCSM_IDLE,
+	SYNCSM_PREP_READ,
+	SYNCSM_FINISH_READ,
+	SYNCSM_PREP_WRITE,
+	SYNCSM_FINISH_WRITE,
+};
+
+struct register_map {
+	u8_t sec;
+	u8_t min;
+	u8_t hour;
+	u8_t dow;
+	u8_t dom;
+	u8_t moncen;
+	u8_t year;
+
+	struct {
+		u8_t sec;
+		u8_t min;
+		u8_t hour;
+		u8_t date;
+	} __packed alarm1;
+
+	struct {
+		u8_t min;
+		u8_t hour;
+		u8_t date;
+	} __packed alarm2;
+
+	u8_t ctrl;
+	u8_t ctrl_stat;
+	u8_t aging;
+	s8_t temp_units;
+	u8_t temp_frac256;
+};
+
+struct gpios {
+	const char *ctrl;
+	gpio_pin_t pin;
+	gpio_dt_flags_t flags;
+};
+
+struct ds3231_config {
+	/* Common structure first because generic API expects this here. */
+	struct counter_config_info generic;
+	const char *bus_name;
+	struct gpios isw_gpios;
+	u16_t addr;
+};
+
+struct ds3231_data {
+	struct device *ds3231;
+	struct device *i2c;
+	struct device *isw;
+	struct register_map registers;
+
+	struct k_sem lock;
+
+	/* Timer structure used for synchronization */
+	struct k_timer sync_timer;
+
+	/* Work structures for the various cases of ISW interrupt. */
+	struct k_work alarm_work;
+	struct k_work sqw_work;
+	struct k_work sync_work;
+
+	/* Forward ISW interrupt to proper worker. */
+	struct gpio_callback isw_callback;
+
+	/* syncclock captured in the last ISW interrupt handler */
+	u32_t isw_syncclock;
+
+	struct rtc_ds3231_syncpoint syncpoint;
+	struct rtc_ds3231_syncpoint new_sp;
+
+	time_t rtc_registers;
+	time_t rtc_base;
+	u32_t syncclock_base;
+
+	/* Pointer to the structure used to notify when a synchronize
+	 * or set operation completes.  Null when nobody's waiting for
+	 * such an operation, or when doing a no-notify synchronize
+	 * through the signal API.
+	 */
+	union {
+		void *ptr;
+		struct sys_notify *notify;
+		struct k_poll_signal *signal;
+	} sync;
+
+	/* Handlers and state when using the counter alarm API. */
+	counter_alarm_callback_t counter_handler[2];
+	u32_t counter_ticks[2];
+
+	/* Handlers and state for DS3231 alarm API. */
+	rtc_ds3231_alarm_callback_handler_t alarm_handler[2];
+	void *alarm_user_data[2];
+	u8_t alarm_flags[2];
+
+	/* Flags recording requests for ISW monitoring. */
+	u8_t isw_mon_req;
+#define ISW_MON_REQ_Alarm 0x01
```

Comment:
That is unusual.

Reply:
You mean mixed case?  I suppose, though it's standard for some roles (e.g. CMSIS-style bitfield values), and is also present internally Zephyr's time_utils.h where the units suffix is (correctly) in lower case.

If it really bothers people I'll uglify it.

Reply:
I meant to define a macro in middle of a struct.